Relations don't work even with default values

**Overview of the problem**

When you add our module but keep all values default, you would expect the app to continue to work. But it doesn't.

With the following configuration :
**Author**
```json
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipName": "hisBook",
            "otherEntityName": "book",
            "relationshipType": "one-to-many",
            "otherEntityRelationshipName": "theAuthor"
        }
    ],
    "fields": [
        {
            "fieldName": "name",
            "fieldType": "String"
        },
        {
            "fieldName": "birthDate",
            "fieldType": "LocalDate"
        }
    ],
    "changelogDate": "20170418140037",
    "dto": "no",
    "service": "no",
    "entityTableName": "author",
    "pagination": "no"
}
```
**Book**
```json
{
    "fluentMethods": true,
    "relationships": [
        {
            "relationshipName": "theAuthor",
            "otherEntityName": "author",
            "relationshipType": "many-to-one",
            "otherEntityField": "name"
        }
    ],
    "fields": [
        {
            "fieldName": "title",
            "fieldType": "String"
        },
        {
            "fieldName": "description",
            "fieldType": "String"
        },
        {
            "fieldName": "publicationDate",
            "fieldType": "LocalDate"
        },
        {
            "fieldName": "price",
            "fieldType": "BigDecimal"
        }
    ],
    "changelogDate": "20170418145015",
    "dto": "no",
    "service": "no",
    "entityTableName": "book",
    "pagination": "no"
}
```

We get the following error message : 
 > Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing column [theAuthor_id] in table [book]

The column is indeed missing because in the following files :
```json
./src/main/resources/config/liquibase/changelog/20170418145015_added_entity_constraints_Book.xml
./src/main/resources/config/liquibase/changelog/20170418145015_added_entity_Book.xml
```
We have `the_author_id` and not `theAuthor_id`. Both are generated from the relationshipName but the former is transformed by jhipster which uses **snake_case** and the latter just appends **_id**. Most probably, JHipster was expecting hibernate to apply SpringNamingStrategy which is not the case thanks to our module.

**Suggest a fix**
We probably will have to find pertinent value in either the ORM or the changelog and make on to copy the other.

**Versions & configurations**

JHipster 4.3.0

--- a/generators/dbh.js
+++ b/generators/dbh.js
@@ -1,7 +1,7 @@
 const DBH_CONSTANTS = require('./dbh-constants');
 const fs = require('fs');
 const jhipsterCore = require('jhipster-core');
-
+const pluralize = require('pluralize');
 
 /** a promise returning the current JHipster app config as a JSON object */
 const getAppConfig = directory => new Promise((resolve, reject) => {
@@ -66,10 +66,58 @@ const validateTableName = (input, dbType) => {
     return true;
 };
 
+/**
+ * get hibernate SnakeCase in JHipster preferred style.
+ *
+ * @param {string} value - table column name or table name string
+ * @see org.springframework.boot.orm.jpa.hibernate.SpringNamingStrategy
+ */
+const hibernateSnakeCase = (value) => {
+    let res = '';
+    if (value) {
+        value = value.replace('.', '_');
+        res = value[0];
+        for (let i = 1, len = value.length - 1; i < len; i++) {
+            if (value[i - 1] !== value[i - 1].toUpperCase() &&
+                value[i] !== value[i].toLowerCase() &&
+                value[i + 1] !== value[i + 1].toUpperCase()
+            ) {
+                res += `_${value[i]}`;
+            } else {
+                res += value[i];
+            }
+        }
+        res += value[value.length - 1];
+        res = res.toLowerCase();
+    }
+    return res;
+};
+
+const getRelationColumn = (name) => hibernateSnakeCase(name) + '_id';
+
+const getPluralRelationColumn = (name) => getRelationColumn(pluralize(name));
+
+const hasConstraints = (relationships) => {
+    for(let idx = 0; idx < relationships.length; idx++) {
+        if(relationships[idx].relationshipType === 'many-to-one' || (relationships[idx].relationshipType === 'one-to-one' && relationships[idx].ownerSide)) {
+            return true;
+        }
+        if(relationships[idx].relationshipType === 'many-to-many' && relationships[idx].ownerSide) {
+            return true;
+        }
+    }
+
+    return false;
+};
+
 
 module.exports = {
     getAppConfig,
     isTrueString,
     validateColumnName,
-    validateTableName
+    validateTableName,
+    getRelationColumn,
+    getPluralRelationColumn,
+    pluralize,
+    hasConstraints
 };

--- a/generators/dbh.js
+++ b/generators/dbh.js
@@ -93,9 +93,9 @@ const hibernateSnakeCase = (value) => {
     return res;
 };
 
-const getRelationColumn = (name) => hibernateSnakeCase(name) + '_id';
+const getColumnIdName = (name) => hibernateSnakeCase(name) + '_id';
 
-const getPluralRelationColumn = (name) => getRelationColumn(pluralize(name));
+const getPluralColumnIdName = (name) => getColumnIdName(pluralize(name));
 
 const hasConstraints = (relationships) => {
     for(let idx = 0; idx < relationships.length; idx++) {
@@ -116,8 +116,7 @@ module.exports = {
     isTrueString,
     validateColumnName,
     validateTableName,
-    getRelationColumn,
-    getPluralRelationColumn,
-    pluralize,
+    getColumnIdName,
+    getPluralColumnIdName,
     hasConstraints
 };

--- a/generators/fix-entity/index.js
+++ b/generators/fix-entity/index.js
@@ -39,7 +39,7 @@ module.exports = generator.extend({
         );
         this.appConfig = jhipsterVar.jhipsterConfig;
 
-        //* / TODO remove on prod
+        /* / TODO remove on prod
         this.prodDatabaseType = jhipsterVar.prodDatabaseType;
         this.log(chalk.blue('<<<<<BEFORE'));
         this.log(chalk.blue('entityConfig'));
@@ -74,14 +74,21 @@ module.exports = generator.extend({
      * Allows consistent mapping with an existing database table without modifying JHipster's entity subgenerator.
      **/
     writing() {
+        /**
+         * Return path to the liquibase file corresponding to this entity and type of file.
+         *
+         * @param type is either 'entity' or 'entity_constraints'
+         */
+        const getLiquibaseFile = (type) => `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_${type}_${this.entityConfig.entityClass}.xml`;
+
         const files = {
             config: this.entityConfig.filename,
             ORM: `${jhipsterVar.javaDir}domain/${this.entityConfig.entityClass}.java`,
-            liquibaseEntity: `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_entity_${this.entityConfig.entityClass}.xml`
+            liquibaseEntity: getLiquibaseFile('entity')
         };
 
         if(dbh.hasConstraints(this.relationships)) {
-            files.liquibaseConstraints = `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_entity_constraints_${this.entityConfig.entityClass}.xml`;
+            files.liquibaseConstraints = getLiquibaseFile('entity_constraints');
         }
 
         // Add/Change/Keep dbhTableName
@@ -137,23 +144,36 @@ module.exports = generator.extend({
             jhipsterFunc.replaceContent(files.liquibaseEntity, `\\<column name="(${columnItem.fieldNameAsDatabaseColumn}|${oldValue})`, `<column name="${newValue}`, true);
         });
 
+        // Add/Change/Keep dbhRelationshipId
         this.relationships.forEach((relationshipItem) => {
-            const oldValue = relationshipItem.dbhRelationName; // todo add it to conf and retrieve it here
+            const pattern = `"relationshipName": "${relationshipItem.relationshipName}"`;
+            const key = 'dbhRelationshipId';
+            const oldValue = relationshipItem.dbhRelationshipId;
+
             let columnName = null;
             let newValue = null;
 
             if(relationshipItem.relationshipType === 'many-to-one' || (relationshipItem.relationshipType === 'one-to-one' && relationshipItem.ownerSide)) {
-                columnName = dbh.getRelationColumn(relationshipItem.relationshipName);
+                columnName = dbh.getColumnIdName(relationshipItem.relationshipName);
                 newValue = relationshipItem.relationshipName + '_id';
             } else if (relationshipItem.relationshipType === 'many-to-many' && relationshipItem.ownerSide) {
-                columnName = dbh.getPluralRelationColumn(relationshipItem.relationshipName);
-                newValue = dbh.pluralize(relationshipItem.relationshipName) + '_id';
-                // todo There are two columns in a many-to-many, need to see what to do with the second one. I think the second one as to do with intermediary table, most probably not my problem
+                columnName = dbh.getPluralColumnIdName(relationshipItem.relationshipName);
+                newValue = relationshipItem.relationshipNamePlural + '_id';
             } else {
+                // If an entity has several relationships but some don't add constraints, they will end up here
                 return;
             }
 
+            if (oldValue === undefined) {
+                // '(\\s*)' is for capturing indentation
+                jhipsterFunc.replaceContent(files.config, `(\\s*)${pattern}`, `$1${pattern},$1"${key}": "${newValue}"`, true);
+            } else {
+                jhipsterFunc.replaceContent(files.config, `"${key}": "${oldValue}`, `"${key}": "${newValue}`);
+            }
+
+            jhipsterFunc.replaceContent(files.ORM, `inverseJoinColumns = @JoinColumn\\(name="(${columnName}|${oldValue})`, `inverseJoinColumns = @JoinColumn(name="${newValue}`, true);
             jhipsterFunc.replaceContent(files.liquibaseEntity, `\\<column name="(${columnName}|${oldValue})`, `<column name="${newValue}`, true);
+            jhipsterFunc.replaceContent(files.liquibaseEntity, `\\<addPrimaryKey columnNames="${dbh.getPluralColumnIdName(this.entityTableName)}, (${columnName}|${oldValue})`, `<addPrimaryKey columnNames="${dbh.getPluralColumnIdName(this.entityTableName)}, ${newValue}`, true);
             jhipsterFunc.replaceContent(files.liquibaseConstraints, `\\<addForeignKeyConstraint baseColumnNames="(${columnName}|${oldValue})`, `<addForeignKeyConstraint baseColumnNames="${newValue}`, true);
         });
     },

--- a/generators/fix-entity/index.js
+++ b/generators/fix-entity/index.js
@@ -2,6 +2,7 @@ const generator = require('yeoman-generator');
 const chalk = require('chalk');
 const prompts = require('./prompts.js');
 const fs = require('fs');
+const dbh = require('../dbh.js');
 
 
 const jhipsterVar = {
@@ -20,6 +21,7 @@ module.exports = generator.extend({
         this.entityTableName = this.options.entityConfig.entityTableName;
         this.dbhTableName = this.options.entityConfig.data.dbhTableName;
         this.fields = this.options.entityConfig.data.fields;
+        this.relationships = this.options.entityConfig.data.relationships;
 
         // input from user (prompts.js will fill them)
         this.tableNameInput = null;
@@ -37,13 +39,15 @@ module.exports = generator.extend({
         );
         this.appConfig = jhipsterVar.jhipsterConfig;
 
-        /* / TODO remove on prod
+        //* / TODO remove on prod
         this.prodDatabaseType = jhipsterVar.prodDatabaseType;
         this.log(chalk.blue('<<<<<BEFORE'));
         this.log(chalk.blue('entityConfig'));
         this.log(this.entityConfig);
         this.log(chalk.blue('fields'));
         this.log(this.fields);
+        this.log(chalk.blue('relations'));
+        this.log(this.options.entityConfig.data.relationships);
         this.log(chalk.blue('jhipsterVar'));
         this.log(jhipsterVar);
         //*/
@@ -73,8 +77,12 @@ module.exports = generator.extend({
         const files = {
             config: this.entityConfig.filename,
             ORM: `${jhipsterVar.javaDir}domain/${this.entityConfig.entityClass}.java`,
-            liquibase: `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_entity_${this.entityConfig.entityClass}.xml`
+            liquibaseEntity: `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_entity_${this.entityConfig.entityClass}.xml`
         };
+
+        if(dbh.hasConstraints(this.relationships)) {
+            files.liquibaseConstraints = `${jhipsterVar.resourceDir}config/liquibase/changelog/${this.entityConfig.data.changelogDate}_added_entity_constraints_${this.entityConfig.entityClass}.xml`;
+        }
 
         // Add/Change/Keep dbhTableName
         const replaceTableName = (paramFiles) => {
@@ -92,7 +100,7 @@ module.exports = generator.extend({
 
             // We search either for our value or jhipster value, so it works even if user didn't accept JHipster overwrite after a regeneration
             jhipsterFunc.replaceContent(paramFiles.ORM, `@Table\\(name = "(${this.entityTableName}|${oldValue})`, `@Table(name = "${newValue}`, true);
-            jhipsterFunc.replaceContent(paramFiles.liquibase, `\\<createTable tableName="(${this.entityTableName}|${oldValue})`, `<createTable tableName="${newValue}`, true);
+            jhipsterFunc.replaceContent(paramFiles.liquibaseEntity, `\\<createTable tableName="(${this.entityTableName}|${oldValue})`, `<createTable tableName="${newValue}`, true);
         };
 
         // DEBUG : log where we are
@@ -105,6 +113,8 @@ module.exports = generator.extend({
                 throw new Error(`JHipster-db-helper : File not found (${file}: ${files[file]}).`);
             }
         }
+
+        this.log(files); // todo remove
 
         replaceTableName(files);
 
@@ -124,7 +134,27 @@ module.exports = generator.extend({
 
             // We search either for our value or jhipster value, so it works even if user didn't accept JHipster overwrite after a regeneration
             jhipsterFunc.replaceContent(files.ORM, `@Column\\(name = "(${columnItem.fieldNameAsDatabaseColumn}|${oldValue})`, `@Column(name = "${newValue}`, true);
-            jhipsterFunc.replaceContent(files.liquibase, `\\<column name="(${columnItem.fieldNameAsDatabaseColumn}|${oldValue})`, `<column name="${newValue}`, true);
+            jhipsterFunc.replaceContent(files.liquibaseEntity, `\\<column name="(${columnItem.fieldNameAsDatabaseColumn}|${oldValue})`, `<column name="${newValue}`, true);
+        });
+
+        this.relationships.forEach((relationshipItem) => {
+            const oldValue = relationshipItem.dbhRelationName; // todo add it to conf and retrieve it here
+            let columnName = null;
+            let newValue = null;
+
+            if(relationshipItem.relationshipType === 'many-to-one' || (relationshipItem.relationshipType === 'one-to-one' && relationshipItem.ownerSide)) {
+                columnName = dbh.getRelationColumn(relationshipItem.relationshipName);
+                newValue = relationshipItem.relationshipName + '_id';
+            } else if (relationshipItem.relationshipType === 'many-to-many' && relationshipItem.ownerSide) {
+                columnName = dbh.getPluralRelationColumn(relationshipItem.relationshipName);
+                newValue = dbh.pluralize(relationshipItem.relationshipName) + '_id';
+                // todo There are two columns in a many-to-many, need to see what to do with the second one. I think the second one as to do with intermediary table, most probably not my problem
+            } else {
+                return;
+            }
+
+            jhipsterFunc.replaceContent(files.liquibaseEntity, `\\<column name="(${columnName}|${oldValue})`, `<column name="${newValue}`, true);
+            jhipsterFunc.replaceContent(files.liquibaseConstraints, `\\<addForeignKeyConstraint baseColumnNames="(${columnName}|${oldValue})`, `<addForeignKeyConstraint baseColumnNames="${newValue}`, true);
         });
     },
 

--- a/integration.md
+++ b/integration.md
@@ -171,9 +171,10 @@ A working application shouldn't print any exception in the logs, nor there shoul
 	* [ ] Modify Book's tableName
 	* [ ] Modify Author's tableName
 * Application with full configuration and module
-	* [ ] Keep all values defaults
+	* [x] Keep all values defaults
 	> Caused by: org.hibernate.tool.schema.spi.SchemaManagementException: Schema-validation: missing column [theAuthor_id] in table [book]
-	[19/04/17 10:00]I don't understand why this fails, I should get the same result that without our module...
+	[19/04/17 10:00] I don't understand why this fails, I should get the same result that without our module...
+	Partly fixed, see issue #3
 	* [ ] Modify Author birthDate's column
 	* [ ] Modify Book any field's column (except the relation)
 	* [ ] Modify Author name's column

